src/S3ClientGUI/mainwindow.cpp

```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
     if (info.isFile() && !m_s3model->getRootBucket().isEmpty()) {
         menu.addAction("Upload", this, SLOT(on_upload()));
         menu.exec(m_fsview->mapToGlobal(point));
+    } else if (info.isDir() && index.data().toString() != ".." && !m_s3model->getRootBucket().isEmpty()) {
+        menu.addAction("Upload", this, SLOT(on_uploadDir()));
+        menu.exec(m_fsview->mapToGlobal(point));
     } else {
         return;
     }
 }
 
+QFileInfoList GetFileList(QString path)
```

Comment:
check this: https://stackoverflow.com/questions/8052460/recursively-iterate-over-all-the-files-in-a-directory-and-its-subdirectories-in

this iterator way is much much better